### DISCUSSION TITLE
Add binary_sensor device_class: heater

### DISCUSSION THREAD

**Discussion by Mariusthvdb** [home-assistant/architecture#841] [Entity Models] [answered]
## Context

The issue motivating this decision, and any context that influences or constrains the decision. Your motivation cannot include that the frontend is not representing the data correctly, as that is a frontend issue.

Suggesting to add a dedicated `device_class: heater` to be able to have the many heating devices adopt the new color scheme in 2022.12, and set a appropriate icon, all default in core HA.

If available in the backend we can set `mdi:radiator` for on, `mdi:radiator-off` and have the colors turn 'red' for on, default off for 'off'

some background: https://github.com/home-assistant/frontend/pull/14637#issuecomment-1348145371

## Proposal

The change that you're proposing.

https://github.com/home-assistant/core/pull/83925

## Consequences

What becomes easier or more difficult to do and any risks introduced by the change that will need to be mitigated.

With this change, we can easily set color vars and have the frontend color that correctly. Without an explicit class, these binaries are colored blue in default schematics, and that is not very appropriate for a Heating device

No risks are involved as far as I can see. Given this is a backend addition required for fixing a frontend issue mainly, there would be no backend issue in the first place, and only advantage in the frontend.

**Comment by MartinHjelmare** [comment]
> Your motivation cannot include that the frontend is not representing the data correctly, as that is a frontend issue.

Please describe what the device class will represent and why it's needed, referencing existing entities eg climate, and why they aren't good enough, instead of talking about icons in the frontend.

**Reply by Mariusthvdb** [reply]
really sorry, but never mind, I have no time for these semantics.

If you're asking why they aren't good enough, you maybe havent followed the latest in https://community.home-assistant.io/t/2022-12-color-states-are-broken-unusable/499890

That rule on 

> Your motivation cannot include that the frontend is not representing the data correctly, as that is a frontend issue.

itself is so completely outdated with 2022.12 changes, I am really starting to wonder why we even bother trying to help the project progress.

take this effort of mine.
1- I asked in Frontend issue/discussion, tells me backend should first discuss, then frontend can fix, so I:
2- asked in devs_core, because backend class is required
3- created a PR in core
4- was told to raise in Architecture, so did
5- was told to instead open discussion in architecture, which again I did
6- now get told this cant be, because Frontend.

really trying to remain civil, and in respect of the Rules of conduct, so Ill leave it with this, and hope someone someday will pickup on this.

thanks.

Ive asked in that thread if someone else will pickup from where I have to leave

**Reply by frenck** [reply]
The main problem is not being civil; the main issue you run into is lacking the developer still needed (don't take this wrong btw, I mean it well). The reason I say this is because, in your path above, you are mixing going to places with your "idea"/"issue", with areas of responsibility in software design itself (the latter is what Martin meant with his comment).

Nevetheless ❤️ for willing to jump in, for real 👍

**Reply by KennethLavrsen** [reply]
You cannot find anywhere in the world where something related to climate - heating or cooling - or both - is not associated with 
- Hot is Red
- Cold is Blue

You look at a watertap. Blue dot for cold water, Red dot for hot water. A device that is made for heating must be shown red when it is heating. An aircon should be blue when cooling. A device that can do both - blue or red.

And off can never be shown as blue or red. I would pick grey.

That is the problem that is attempted solved. If this require new device_class or can be solved by other means is another story. But that is the problem. We have people with Heaters that show Blue when they are ON. That is functionally wrong. It is not a matter of taste. Blue is cold or cooling in all cultures I know.

**Reply by Mariusthvdb** [reply]
ok one last try on my behalf.

Sure I am no developer, never stated I was, and you are correct I asked first (didn't start this discussion out of my own initiative, but was asked to do so)

I talked to the frontend dev, who told me backend should tackle this first. So I asked backend.  etc etc

Someone in that chain of responsibility could/should have said: ok thanks for your idea, we'll take it from here, you are correct we have an issue to solve. Or, thanks for your suggestion, we wont be taking this up.

because now you are teaching me the design principles of software design, which is a waste of time on your behalf, and not solving the problem the frontend has introduced, which is a waste of resources on the communities side of things.

If you really appreciate the efforts, please think along a bit more, and take it up from where this amateur has left it.

**Reply by frenck** [reply]
> ok thanks for your idea

You are not raising an idea, you started making code change and submitting architectural discussions to make a change happen. That is not suggesting an idea, that is building. Those are different things. You are now commenting on the review process of code, while your intention is raising an idea?

PRs and request architectural approval for you implementing a change, is not what an "idea" entails :)

../Frenck

**Reply by MartinHjelmare** [reply]
The feature is still not described. Eg, if this entity should describe the on/off state of a heater, why isn't a climate entity used? Please describe the feature without talking about colors.

**Reply by Mariusthvdb** [reply]
Martin, thanks, you're pulling me back into this ;-) thanks for that.

this is not about climate entities, but about binary_sensor entities, like the ones created by Plugwise core integration:
`binary_sensor.opentherm_dhw_state`, `binary_sensor.opentherm_flame_state` and `binary_sensor.opentherm_heating`

All three have a different meaning, and are essential to our systems logics and automations based on that. The heater can be off, while the dhw_state can be on, or vv.

Not really sure how we can describe the issue better than above. it is an extra pressing issue, because even the climate entities have issues which should be handled separately.

On the suggestion of device_class itself, isn't the concept of a device_class introduced specifically [to allow correct frontend representation](https://www.home-assistant.io/integrations/binary_sensor/#device-class)? In fact, isnt that all it does? So I am not sure how to proceed describing the issue, without mentioning the Frontend here and getting overruled

**Reply by MartinHjelmare** [reply]
The frontend representation isn't important when we talk about entity state features in the backend and want to decide if we should add them or not. The frontend is just one consumer of entity state. The entity state feature needs to make sense on its own.

Are there other integrations that integrate heater devices with read only binary state features, similar to Plugwise? Please search among existing and potential future integrations and make a summary with the number of integrations that could benefit from the new device class.

Before we add new device classes we want to see that more than a single integration would benefit. Normally a handful of integrations are required.

**Reply by mdegat01** [reply]
> On the suggestion of device_class itself, isn't the concept of a device_class introduced specifically [to allow correct frontend representation](https://www.home-assistant.io/integrations/binary_sensor/#device-class)? In fact, isnt that all it does?

No. Although possibly in binary entities. But in sensors it does significantly more. Take `timestamp` for instance. When setting the state of a timestamp type sensor (like in a template sensor) it actually accepts a python datetime object as the state. That's not true for any sensor with a different device class, it only makes sense for timestamp.

Also sensors with a device class of `temperature` can be converted from Celsius to Fahrenheit and vice versa. Again, only for that device class and only because HA knows what it is.

In addition nearly all of them have expected unit of measurement options and a schema for is a valid state value for that device class. Others might have other special features I'm forgetting about as well.

Binary sensors are significantly simpler for now. I'm not sure any device class currently conveys additional functionality or features other then presentation. That doesn't mean it won't in the future though. Plus each one added means existing integrations should be refactoring things to more accurately represent their entities. Hence martins request for examples. Where are entities being incorrectly represented as something else and how would this `heater` option help?

**Reply by Mariusthvdb** [reply]
Mike, I see that, and I was indeed talking device_class for binary_sensor here. no misunderstanding.

Please also chime in @piitaya because it was discussing with Paul this whole evolution started. As FE dev he will be better equipped phrasing things correctly.

meanwhile Ive [asked the community](https://community.home-assistant.io/t/2022-12-color-states-are-broken-unusable/499890/400?u=mariusthvdb) to help me out, while I will search the  existing integrations (not sure how to search potential future integrations) and summarize that when I am back..

**Reply by Kars-de-Jong** [reply]
The custom integration [Luxtronik 2](https://github.com/BenPru/luxtronik) has an indication for the compressor heating being active: `binary_sensor.luxtronik2_compressor_heater`.
I believe the custom integration [Ramses CC](https://github.com/zxdavb/ramses_cc) also has at least one binary sensors that represent a heater being active.

**Reply by Petro31** [reply]
@Kars-de-Jong custom integrations are not accounted for with core.  We are asking for core-integrations that benefit from this device class.

**Reply by Mariusthvdb** [reply]
is there an easy way of searching in climate/heat integration for binary_sensor? what would the search syntax be?

**Comment by pimw1** [comment]
Potential future integrations were also mentioned by Martin. Custom integrations are perfect examples of that.

**Comment by Mariusthvdb** [comment]
https://github.com/home-assistant/core/blob/0e2ebfe5c45716250280186234123f170e3bd08c/homeassistant/components/plugwise/binary_sensor.py#L45

this might even be falsely classed as heat (should be heater?): 
https://github.com/home-assistant/core/blob/b97ec8125f8236a18435c12d7e0abca5e916e8ee/homeassistant/components/overkiz/binary_sensor.py#L104

https://github.com/home-assistant/core/blob/b97ec8125f8236a18435c12d7e0abca5e916e8ee/homeassistant/components/toon/binary_sensor.py#L130

https://github.com/home-assistant/core/blob/b97ec8125f8236a18435c12d7e0abca5e916e8ee/homeassistant/components/vicare/binary_sensor.py#L53

and several other binary_sensors, like the Plugwise in Vicare

this would be an example of future integration being set to 'heater'
https://github.com/home-assistant/core/blob/b97ec8125f8236a18435c12d7e0abca5e916e8ee/homeassistant/components/incomfort/water_heater.py#L67

and

https://github.com/home-assistant/core/blob/b97ec8125f8236a18435c12d7e0abca5e916e8ee/homeassistant/components/atag/water_heater.py#L40

https://github.com/home-assistant/core/blob/b97ec8125f8236a18435c12d7e0abca5e916e8ee/homeassistant/components/intellifire/binary_sensor.py#L38

not sure about this NIbe heat pump:
https://github.com/home-assistant/core/blob/b97ec8125f8236a18435c12d7e0abca5e916e8ee/homeassistant/components/nibe_heatpump/binary_sensor.py#L40

Opentherm would be obvious (same as Plugwise, also Opentherm)
https://github.com/home-assistant/core/blob/b97ec8125f8236a18435c12d7e0abca5e916e8ee/homeassistant/components/opentherm_gw/binary_sensor.py#L157

**Reply by MartinHjelmare** [reply]
Seems like enough integrations could leverage the new device class. We need some more affirmations from members before going ahead.

**Comment by ChristophCaina** [comment]
while not using such a component with a core integration, I do have some switches that turn on a heating device in my car (remotely) and I would really love to see such a feature..

[EDIT] - just saw, that this is for binary_sensors...
my recommendation would be for switches ... hope, that this would still count in ? ^^

![grafik](https://user-images.githubusercontent.com/26391061/207660351-fa66d62c-3c35-4f2a-889d-58d78e0a7d6d.png)

**Comment by frenck** [comment, marked as answer]
We've discussed this proposal in our architectual meeting, and have decided against it. We do not see enough cases to support this right now.

Nevertheless, thanks for your proposal.

../Frenck
